spec/ParseRole.spec.js

```diff
@@ -86,6 +87,85 @@ describe('Parse Role testing', () => {
     return acl
   }
 
+  it("should not recursively load the same role multiple times", (done) => {
+    const rootRole = "RootRole";
+    const roleNames = ["FooRole", "BarRole", "BazRole"];
+    const allRoles = [rootRole].concat(roleNames);
+
+    const roleObjs = {};
+    const createAllRoles = function(user) {
+      const promises = allRoles.map(function(roleName) {
+        return createRole(roleName, null, user)
+          .then(function(roleObj) {
+            roleObjs[roleName] = roleObj;
+            return roleObj;
+          });
+      });
+      return Promise.all(promises);
+    };
+
+    const restExecute = spyOn(RestQuery.prototype, "execute").and.callThrough();
+
+    let user,
+      auth,
+      getAllRolesSpy;
+    createTestUser().then((newUser) => {
+      user = newUser;
+      return createAllRoles(user);
+    }).then ((roles) => {
+      const rootRoleObj = roleObjs[rootRole];
+      roles.forEach(function(role, i) {
+        // Add all roles to the RootRole
+        if (role.id !== rootRoleObj.id) {
+          role.relation("roles").add(rootRoleObj);
+        }
+        // Add all "roleNames" roles to the previous role
+        if (i > 0) {
+          role.relation("roles").add(roles[i - 1]);
+        }
+      });
+
+      // create some additional relations between children roles
+      const FooRole = roles.find(function(role) {
+        return role.get("name") === "FooRole"
+      });
+      const BarRole = roles.find(function(role) {
+        return role.get("name") === "BarRole"
+      });
+      const BazRole = roles.find(function(role) {
+        return role.get("name") === "BazRole"
+      });
+      BarRole.relation("roles").add([FooRole, BazRole]);
+      BazRole.relation("roles").add([BarRole, FooRole]);
+
+      return Parse.Object.saveAll(roles, { useMasterKey: true });
+    }).then(() => {
+      auth = new Auth({config: Config.get("test"), isMaster: true, user: user});
+      const authRoles = auth.getAuthRoles();
+      getAllRolesSpy = spyOn(authRoles, "_findAndBuildRolesForRolesRecursivelyOntoMap").and.callThrough();
+
+      return authRoles.findRoles();
+    }).then ((roles) => {
+      expect(roles.length).toEqual(4);
+
+      allRoles.forEach(function(name) {
+        expect(roles.indexOf("role:" + name)).not.toBe(-1);
+      });
+
+      // 1 query for the direct roles (parent roles).
+      // 1 query for each role after that, including the parent role.
+      expect(restExecute.calls.count()).toEqual(5);
```

Comment:
@flovilmart  An increase in calls is to be expected here since for each role we have to know the parent to properly compile the tree.